evm-contracts/src/v0.6/Median.sol

```diff
@@ -4,105 +4,233 @@ import "./vendor/SafeMath.sol";
 import "./vendor/SignedSafeMath.sol";
 
 library Median {
-  using SafeMath for uint256;
   using SignedSafeMath for int256;
 
-  /**
-   * @dev Returns the sorted middle, or the average of the two middle indexed 
-   * items if the array has an even number of elements
-   * @param _list The list of elements to compare
-   */
-  function calculate(int256[] memory _list)
+  /// @notice Returns the sorted middle, or the average of the two middle indexed items if the
+  /// array has an even number of elements.
+  /// @dev The list passed as an argument isn't modified.
+  /// @dev This algorithm has expected runtime O(n), but for adversarially chosen inputs
+  /// the runtime is O(n^2).
+  /// @param list The list of elements to compare
+  function calculate(int256[] memory list)
     internal
     pure
     returns (int256)
   {
-    uint256 answerLength = _list.length;
-    uint256 middleIndex = answerLength.div(2);
-    if (answerLength % 2 == 0) {
-      int256 median1 = quickselect(copy(_list), middleIndex);
-      int256 median2 = quickselect(_list, middleIndex.add(1)); // quickselect is 1 indexed
-      int256 remainder = (median1 % 2 + median2 % 2) / 2;
-      return (median1 / 2).add(median2 / 2).add(remainder); // signed integers are not supported by SafeMath
+    require(0 < list.length, "list must not be empty");
```

Comment:
Since we're using the same require in multiple places, I'd suggest making a modifier for it. Something like `onlyValidLength`

Reply:
Since it's only in two places, and the complexity of the check is low, I'd prefer to keep as is for better readability.

Reply:
Good with me.